### ISSUE TITLE
Lint for duplicate impl section

### ISSUE THREAD

**Issue by mspiegel** [rust-lang/rust-clippy#1024]
I'm a novice to rust and noticed recently that I had accidentally created two implementation sections within the same module. Both sections were impl's that did not have a trait declaration:

```
impl Foobar {
   ...stuff here
}

impl Foobar {
   ...more stuff here
}
```

It's probably a good practice to warn on duplicate impls or is there something I am not considering?

**Comment by Manishearth** [comment]
Duplicate impls aren't really an antipattern, but this could be an `allow` lint

**Comment by llogiq** [comment]
I see duplicate impls in conjunction with `#[cfg(..)]`s to unify implementations for different systems. Otherwise I consider them a net loss in readability.

**Comment by mcarton** [comment]
I also like to (ab?)use that to make sections in the documentation:

```rust
pub struct Foo;

/// # Constructors
impl Foo {
    /// Make a default `Foo`
    pub fn new() -> Foo { … }

    /// Make a `Foo` with some `Bar`
    pub fn with_bar() -> Foo { … }
}

/// # Operations on `Foo`s
impl Foo {
    /// Do something with a `Foo`
    pub fn do_sth(self) { … }
}
```

looks like this:

![screen shot 2016-06-19 at 13 28 16](https://cloud.githubusercontent.com/assets/3751788/16177041/e26fe6f0-3621-11e6-91ca-f8ea6e05f6d9.png)

But:

```rust
pub struct Foo;

impl Foo {
    /// # Constructors
    /// Make a default `Foo`
    pub fn new() -> Foo { … }

    /// Make a `Foo` with some `Bar`
    pub fn with_bar() -> Foo { … }

    /// # Operations on `Foo`s
    /// Do something with a `Foo`
    pub fn do_sth(self) { … }
}
```

looks like this:

![screen shot 2016-06-19 at 13 31 27](https://cloud.githubusercontent.com/assets/3751788/16177051/3b4457ac-3622-11e6-9199-bebd583717c0.png)

Ok for an allow-by-default lint though.

**Comment by camsteffen** [comment]
Duplicates #414
